src/Tracer.ts

```diff
@@ -1,297 +1,79 @@
-    let updatedEvent = this.preProcess(event)
-    let result = await this.client.log(updatedEvent)
-    return this.postProcess(result)
-  }
-
-
-  private async logWithAction(message: string | { [key: string]: NonNullable<any> }, loggerOptions: LoggerOptions) {
-    if (!message) throw new Error('no message to provided')
-    if (this._traceContext.finishTimestamp) throw new Error('span finished. no further actions allowed')
-    let { state, action = LogEventAction.info } = extractLoggerOptions(EventType.log, loggerOptions)
-    let messageToLog
-    if (!state) throw new Error('no valid state provided')
-    if (typeof message === 'string') {
-      messageToLog = new EventMessage({
-        content: {
-          message
-        },
-        type: 'log'
-      })
-    } else if ((typeof message === 'object') && (!(message.hasOwnProperty('content')) || !(message.hasOwnProperty('type')))) {
-      messageToLog = new EventMessage({
-        content: message,
-        type: 'log'
-      })
-    } else {
-      messageToLog = new EventMessage(<IEventMessage>message)
-    }
-    let newEnvelope = Object.assign(messageToLog, {
-      metadata: {
-        event: EventMetadata.log({
-          action,
-          state
-        }),
-        trace: <IEventTrace>this._traceContext
-      }
-    })
-
-    let logResult = await this.record(newEnvelope);
-    if (LogResponseStatus.accepted == logResult.status) {
-      return logResult
-    } else {
-      throw new Error(`Error when logging trace. status: ${logResult.status}`)
-    }
-  }
-}
-
-const buildLoggerOptions = (actionDefault: EventAction, action: EventAction, state?: EventStateMetadata): LoggerOptions => {
-  let result: LoggerOptions = {
-    action, state
-  }
-  result.action = action ? action : actionDefault
-  result.state = state ? state : EventStateMetadata.success()
-  return result
-}
-
-const extractLoggerOptions = (type: EventType, loggerOptions: LoggerOptions): LoggerOptions => {
-  let { action = NullEventAction.undefined, state } = loggerOptions
-  switch (type) {
-    case EventType.audit: {
-      return buildLoggerOptions(AuditEventAction.default, action, state)
-    }
-    case EventType.trace: {
-      return buildLoggerOptions(TraceEventAction.span, action, state)
-    }
-    case EventType.log: {
-      return buildLoggerOptions(LogEventAction.info, action, state)
-    }
-    default: {
-      return buildLoggerOptions(NullEventAction.undefined, NullEventAction.undefined)
+  static extractContextFromMessage(message: { [key: string]: any }, extractOptions: ContextOptions = {}): TypeSpanContext {
+    let spanContext
+    let { path } = extractOptions // type not implemented yet
+    if (message instanceof EventMessage || (('metadata' in message) && 'trace' in message.metadata)) {
+      path = 'metadata.trace'
+    } else if ('trace' in message) {
+      path = 'trace'
     }
+    spanContext = new EventTraceMetadata(<TypeSpanContext>getNestedObject(message, path))
```

Comment:
Please see my comments about using lodash (or at least wrap the lodash operations in the `getNestedOjbect` util function.